// @solana/buffer-layout-utils/lib/cjs/bigint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.u256be = exports.u256 = exports.u192be = exports.u192 = exports.u128be = exports.u128 = exports.u64be = exports.u64 = exports.bigIntBE = exports.bigInt = void 0;
const buffer_1 = require("buffer");
const buffer_layout_1 = require("@solana/buffer-layout");
const bigint_buffer_1 = require("bigint-buffer");
const base_1 = require("./base");
// https://github.com/no2chem/bigint-buffer/issues/59
function assertValidBigInteger(untrustedInput) {
    if (typeof untrustedInput !== 'bigint') {
        throw new Error('Expected a `BigInt`');
    }
}
function bigInt_IMPL(littleEndian, length) {
    return (property) => {
        const layout = (0, buffer_layout_1.blob)(length, property);
        const { encode, decode } = (0, base_1.encodeDecode)(layout);
        const bigIntLayout = layout;
        bigIntLayout.decode = (buffer, offset) => {
            const src = decode(buffer, offset);
            return littleEndian ? (0, bigint_buffer_1.toBigIntLE)(buffer_1.Buffer.from(src)) : (0, bigint_buffer_1.toBigIntBE)(buffer_1.Buffer.from(src));
        };
        bigIntLayout.encode = (bigInt, buffer, offset) => {
            assertValidBigInteger(bigInt);
            let src;
            if (length === 0) {
                // https://github.com/no2chem/bigint-buffer/issues/40
                // `toBuffer{BE|LE}` crashes when passed zero for the `length` argument.
                src = buffer_1.Buffer.alloc(0);
            }
            else {
                src = littleEndian ? (0, bigint_buffer_1.toBufferLE)(bigInt, length) : (0, bigint_buffer_1.toBufferBE)(bigInt, length);
            }
            return encode(src, buffer, offset);
        };
        return bigIntLayout;
    };
}
exports.bigInt = bigInt_IMPL.bind(null, /* littleEndian */ true);
exports.bigIntBE = bigInt_IMPL.bind(null, /* littleEndian */ false);
exports.u64 = (0, exports.bigInt)(8);
exports.u64be = (0, exports.bigIntBE)(8);
exports.u128 = (0, exports.bigInt)(16);
exports.u128be = (0, exports.bigIntBE)(16);
exports.u192 = (0, exports.bigInt)(24);
exports.u192be = (0, exports.bigIntBE)(24);
exports.u256 = (0, exports.bigInt)(32);
exports.u256be = (0, exports.bigIntBE)(32);
