// @solana/spl-token/lib/cjs/extensions/permissionedBurn/instructions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PermissionedBurnInstruction = void 0;
exports.createInitializePermissionedBurnInstruction = createInitializePermissionedBurnInstruction;
exports.createPermissionedBurnInstruction = createPermissionedBurnInstruction;
exports.createPermissionedBurnCheckedInstruction = createPermissionedBurnCheckedInstruction;
const buffer_layout_1 = require("@solana/buffer-layout");
const buffer_layout_utils_1 = require("@solana/buffer-layout-utils");
const web3_js_1 = require("@solana/web3.js");
const constants_js_1 = require("../../constants.js");
const errors_js_1 = require("../../errors.js");
const internal_js_1 = require("../../instructions/internal.js");
const types_js_1 = require("../../instructions/types.js");
var PermissionedBurnInstruction;
(function (PermissionedBurnInstruction) {
    PermissionedBurnInstruction[PermissionedBurnInstruction["Initialize"] = 0] = "Initialize";
    PermissionedBurnInstruction[PermissionedBurnInstruction["Burn"] = 1] = "Burn";
    PermissionedBurnInstruction[PermissionedBurnInstruction["BurnChecked"] = 2] = "BurnChecked";
})(PermissionedBurnInstruction || (exports.PermissionedBurnInstruction = PermissionedBurnInstruction = {}));
const initializePermissionedBurnInstructionData = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)('instruction'),
    (0, buffer_layout_1.u8)('permissionedBurnInstruction'),
    (0, buffer_layout_utils_1.publicKey)('authority'),
]);
/**
 * Construct a InitializePermissionedBurnConfig instruction
 *
 * @param mint          Token mint account
 * @param authority     The permissioned burn mint's authority
 * @param programId     SPL Token program account
 */
function createInitializePermissionedBurnInstruction(mint, authority, programId = constants_js_1.TOKEN_2022_PROGRAM_ID) {
    if (!(0, constants_js_1.programSupportsExtensions)(programId)) {
        throw new errors_js_1.TokenUnsupportedInstructionError();
    }
    const keys = [{ pubkey: mint, isSigner: false, isWritable: true }];
    const data = Buffer.alloc(initializePermissionedBurnInstructionData.span);
    initializePermissionedBurnInstructionData.encode({
        instruction: types_js_1.TokenInstruction.PermissionedBurnExtension,
        permissionedBurnInstruction: PermissionedBurnInstruction.Initialize,
        authority,
    }, data);
    return new web3_js_1.TransactionInstruction({ keys, programId, data });
}
const permissionedBurnInstructionData = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)('instruction'),
    (0, buffer_layout_1.u8)('permissionedBurnInstruction'),
    (0, buffer_layout_utils_1.u64)('amount'),
]);
/**
 * Construct a permissioned burn instruction
 *
 * @param account                       Token account to update
 * @param mint                          Token mint account
 * @param owner                         The account's owner/delegate
 * @param permissionedBurnAuthority     Authority configured on the mint for permissioned burns
 * @param amount                        Amount to burn
 * @param multiSigners                  The signer account(s)
 * @param programId                     SPL Token program account
 */
function createPermissionedBurnInstruction(account, mint, owner, permissionedBurnAuthority, amount, multiSigners = [], programId = constants_js_1.TOKEN_2022_PROGRAM_ID) {
    if (!(0, constants_js_1.programSupportsExtensions)(programId)) {
        throw new errors_js_1.TokenUnsupportedInstructionError();
    }
    const keys = (0, internal_js_1.addSigners)([
        { pubkey: account, isSigner: false, isWritable: true },
        { pubkey: mint, isSigner: false, isWritable: true },
        { pubkey: permissionedBurnAuthority, isSigner: true, isWritable: false },
    ], owner, multiSigners);
    const data = Buffer.alloc(permissionedBurnInstructionData.span);
    permissionedBurnInstructionData.encode({
        instruction: types_js_1.TokenInstruction.PermissionedBurnExtension,
        permissionedBurnInstruction: PermissionedBurnInstruction.Burn,
        amount: BigInt(amount),
    }, data);
    return new web3_js_1.TransactionInstruction({ keys, programId, data });
}
const permissionedBurnCheckedInstructionData = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)('instruction'),
    (0, buffer_layout_1.u8)('permissionedBurnInstruction'),
    (0, buffer_layout_utils_1.u64)('amount'),
    (0, buffer_layout_1.u8)('decimals'),
]);
/**
 * Construct a checked permissioned burn instruction
 *
 * @param account                       Token account to update
 * @param mint                          Token mint account
 * @param owner                         The account's owner/delegate
 * @param permissionedBurnAuthority     Authority configured on the mint for permissioned burns
 * @param amount                        Amount to burn
 * @param decimals                      Number of the decimals of the mint
 * @param multiSigners                  The signer account(s)
 * @param programId                     SPL Token program account
 */
function createPermissionedBurnCheckedInstruction(account, mint, owner, permissionedBurnAuthority, amount, decimals, multiSigners = [], programId = constants_js_1.TOKEN_2022_PROGRAM_ID) {
    if (!(0, constants_js_1.programSupportsExtensions)(programId)) {
        throw new errors_js_1.TokenUnsupportedInstructionError();
    }
    const keys = (0, internal_js_1.addSigners)([
        { pubkey: account, isSigner: false, isWritable: true },
        { pubkey: mint, isSigner: false, isWritable: true },
        { pubkey: permissionedBurnAuthority, isSigner: true, isWritable: false },
    ], owner, multiSigners);
    const data = Buffer.alloc(permissionedBurnCheckedInstructionData.span);
    permissionedBurnCheckedInstructionData.encode({
        instruction: types_js_1.TokenInstruction.PermissionedBurnExtension,
        permissionedBurnInstruction: PermissionedBurnInstruction.BurnChecked,
        amount: BigInt(amount),
        decimals,
    }, data);
    return new web3_js_1.TransactionInstruction({ keys, programId, data });
}
