// @solana/spl-token/lib/cjs/serialization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COptionU64Layout = exports.COptionPublicKeyLayout = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
const buffer_layout_utils_1 = require("@solana/buffer-layout-utils");
class COptionPublicKeyLayout extends buffer_layout_1.Layout {
    constructor(property) {
        super(-1, property);
        this.publicKeyLayout = (0, buffer_layout_utils_1.publicKey)();
    }
    decode(buffer, offset = 0) {
        const option = buffer[offset];
        if (option === 0) {
            return null;
        }
        return this.publicKeyLayout.decode(buffer, offset + 1);
    }
    encode(src, buffer, offset = 0) {
        if (src === null) {
            buffer[offset] = 0;
            return 1;
        }
        else {
            buffer[offset] = 1;
            this.publicKeyLayout.encode(src, buffer, offset + 1);
            return 33;
        }
    }
    getSpan(buffer, offset = 0) {
        if (buffer) {
            const option = buffer[offset];
            return option === 0 ? 1 : 1 + this.publicKeyLayout.span;
        }
        throw new RangeError('Buffer must be provided');
    }
}
exports.COptionPublicKeyLayout = COptionPublicKeyLayout;
class COptionU64Layout extends buffer_layout_1.Layout {
    constructor(property) {
        super(-1, property);
        this.u64Layout = (0, buffer_layout_utils_1.u64)();
    }
    decode(buffer, offset = 0) {
        const option = buffer[offset];
        if (option === 0) {
            return null;
        }
        return this.u64Layout.decode(buffer, offset + 1);
    }
    encode(src, buffer, offset = 0) {
        if (src === null) {
            buffer[offset] = 0;
            return 1;
        }
        else {
            buffer[offset] = 1;
            this.u64Layout.encode(src, buffer, offset + 1);
            return 9;
        }
    }
    getSpan(buffer, offset = 0) {
        if (buffer) {
            const option = buffer[offset];
            return option === 0 ? 1 : 1 + this.u64Layout.span;
        }
        throw new RangeError('Buffer must be provided');
    }
}
exports.COptionU64Layout = COptionU64Layout;
