// @solana/spl-token/lib/cjs/instructions/unwrapLamports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unwrapLamportsInstructionData = void 0;
exports.createUnwrapLamportsInstruction = createUnwrapLamportsInstruction;
exports.decodeUnwrapLamportsInstruction = decodeUnwrapLamportsInstruction;
exports.decodeUnwrapLamportsInstructionUnchecked = decodeUnwrapLamportsInstructionUnchecked;
const buffer_layout_1 = require("@solana/buffer-layout");
const web3_js_1 = require("@solana/web3.js");
const errors_js_1 = require("../errors.js");
const internal_js_1 = require("./internal.js");
const types_js_1 = require("./types.js");
const serialization_js_1 = require("../serialization.js");
/** TODO: docs */
exports.unwrapLamportsInstructionData = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)('instruction'),
    new serialization_js_1.COptionU64Layout('amount'),
]);
/**
 * Construct a UnwrapLamports instruction
 *
 * @param source       Native source account
 * @param destination  Account receiving the lamports
 * @param owner        Owner of the source account
 * @param amount       Amount of lamports to unwrap
 * @param multiSigners Signing accounts if `owner` is a multisig
 * @param programId    SPL Token program account
 *
 * @return Instruction to add to a transaction
 */
function createUnwrapLamportsInstruction(source, destination, owner, amount, multiSigners = [], programId) {
    const keys = (0, internal_js_1.addSigners)([
        { pubkey: source, isSigner: false, isWritable: true },
        { pubkey: destination, isSigner: false, isWritable: true },
    ], owner, multiSigners);
    const data = Buffer.alloc(10); // worst-case
    exports.unwrapLamportsInstructionData.encode({
        instruction: types_js_1.TokenInstruction.UnwrapLamports,
        amount,
    }, data);
    return new web3_js_1.TransactionInstruction({ keys, programId, data });
}
/**
 * Decode a UnwrapLamports instruction and validate it
 *
 * @param instruction Transaction instruction to decode
 * @param programId   SPL Token program account
 *
 * @return Decoded, valid instruction
 */
function decodeUnwrapLamportsInstruction(instruction, programId) {
    if (!instruction.programId.equals(programId))
        throw new errors_js_1.TokenInvalidInstructionProgramError();
    if (instruction.data.length < exports.unwrapLamportsInstructionData.getSpan(instruction.data))
        throw new errors_js_1.TokenInvalidInstructionDataError();
    const { keys: { source, destination, owner, multiSigners }, data, } = decodeUnwrapLamportsInstructionUnchecked(instruction);
    if (data.instruction !== types_js_1.TokenInstruction.UnwrapLamports)
        throw new errors_js_1.TokenInvalidInstructionTypeError();
    if (!source || !destination || !owner)
        throw new errors_js_1.TokenInvalidInstructionKeysError();
    return {
        programId,
        keys: {
            source,
            destination,
            owner,
            multiSigners,
        },
        data,
    };
}
/**
 * Decode a UnwrapLamports instruction without validating it
 *
 * @param instruction Transaction instruction to decode
 *
 * @return Decoded, non-validated instruction
 */
function decodeUnwrapLamportsInstructionUnchecked({ programId, keys: [source, destination, owner, ...multiSigners], data, }) {
    return {
        programId,
        keys: {
            source,
            destination,
            owner,
            multiSigners,
        },
        data: exports.unwrapLamportsInstructionData.decode(data),
    };
}
